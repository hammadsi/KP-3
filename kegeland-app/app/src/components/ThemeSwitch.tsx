import React from 'react';
import {Switch, useTheme} from 'react-native-paper';

import useAppDispatch from '~hooks/useAppDispatch';
import useAppSelector from '~hooks/useAppSelector';
import {updateSetting} from '~state/ducks/app/app.actions';

export type ThemeSwitchProps = {
  style?:
    | {
        marginRight: number;
        marginVertical?: number | undefined;
      }
    | undefined;
};

/**
 * Component for switching the app's color theme
 * @param props the props
 * @see {@link ThemeSwitchProps}
 * @see {@link Switch}
 * @see {@link useTheme}
 */
const ThemeSwitch: React.FC<ThemeSwitchProps> = (props) => {
  const {colors} = useTheme();
  const dispatch = useAppDispatch();
  const {darkMode} = useAppSelector((state) => state.app.settings);
  const toggle = () => {
    dispatch(updateSetting<'darkMode'>({key: 'darkMode', value: !darkMode}));
  };
  return (
    <Switch
      color={colors.primary}
      trackColor={{
        false: colors.placeholder,
        true: colors.primary,
      }}
      value={darkMode}
      onValueChange={() => toggle()}
      {...props}
    />
  );
};

export default ThemeSwitch;
