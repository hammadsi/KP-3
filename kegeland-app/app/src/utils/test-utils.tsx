/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
