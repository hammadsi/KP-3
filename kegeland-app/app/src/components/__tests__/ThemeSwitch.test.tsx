import React from 'react';

import {render} from '~utils/test-utils';

import ThemeSwitch from '../ThemeSwitch';

describe('Test ThemeSwitch-component', () => {
  const errorSpy = jest
    .spyOn(global.console, 'error')
    .mockImplementation(() => {});
  afterAll(() => {
    errorSpy.mockRestore();
  });
  it('should render correctly', () => {
    const component = <ThemeSwitch />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
