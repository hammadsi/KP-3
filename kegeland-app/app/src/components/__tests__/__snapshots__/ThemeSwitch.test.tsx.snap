// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ThemeSwitch-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTSwitch
    accessibilityRole="switch"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onTintColor="#00b0ff"
    style={
      Object {
        "height": 31,
        "width": 51,
      }
    }
    tintColor="rgba(0, 0, 0, 0.54)"
    value={false}
  />
</View>
`;
